### DISCUSSION TITLE
Prepbufr Dump File Question

### DISCUSSION THREAD

**Discussion by craigrj** [dtcenter/METplus#1395] [File I/O] [answered]
Where can I found out the code values that appear in the PB2NC ADPSFC dump file.  Specifically, what are ob, qm, pc, rc, fc, an, and cat column headings.  I am looking for 6hr Precip values (TP06).   Where would I find those in the dump file?

What I am trying to do is find out what the pb_report_type, in_report_type, instrumentation type to generate ob files that don't contain Department of transportation precipitation obs for the TP06 variable.  I guess I have to include the types I want to keep.  In the documentation, it says to check the table 4 an 6 in the prepbufr doc but they don't seem to have the information I am after.  Is there a better reference.

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hi @craigrj, yes, good questions. Unfortunately, the MET and METplus User's Guides don't contain that level of detail about individual observation data sources. The relatively new METplus [Verification Datasets Guide](https://metplus.readthedocs.io/en/latest/Verification_Datasets/index.html) is the place where data source-specific details could be added. However, I know that it does not contain the answer to your specific question.

This is the only documentation I'm aware of which describes the contents of PrepBUFR files:
https://www.emc.ncep.noaa.gov/mmb/data_processing/prepbufr.doc/document.htm

Here are direct links to [table4](https://www.emc.ncep.noaa.gov/mmb/data_processing/prepbufr.doc/table_4.htm) and [table6](https://www.emc.ncep.noaa.gov/mmb/data_processing/prepbufr.doc/table_6.htm), but I assume you've already seen these.

I did some testing. First, using the pb2nc -index option with a NAM PrepBUFR file:
```
pb2nc ndas/nam.20120410.t12z.prepbufr.tm00.nr out.nc PB2NCConfig_default -index  -log run_pb2nc.log
```

And the log output contains:
```
DEBUG 1:     TP06: TOTAL PRECIPITATION PAST 6 HOURS                       	ADPSFC
```

So presumably it has TP06 in the ADPSFC message type. Rerunning using the "-dump" option instead of "-index", I see many of the following types of records:
```
07040       1.10  49.93  -1.000     38.0   181.0   511.0 ******
--------------------------------------------------------------------------------
lev    1       ob       qm       pc       rc       fc       an      cat
--------------------------------------------------------------------------------
       P    988.6      2.0      1.0                                 0.0
       Q   6070.0      2.0      8.0      0.0                        0.0
       Q   6070.0      2.0      1.0
       T     10.7      2.0      8.0      0.0                        0.0
       T      9.7      2.0      1.0
       Z     38.0      2.0      1.0                                 0.0
```

Unfortunately, this is NOT what you're looking for! The dumppb.f code in pb2nc only dumps the event array that contains observations for P, Q, T, Z, U, and V (pressure, specific humidity, temp, height, u-wind, and v-wind, respectively). See [this line](https://github.com/dtcenter/MET/blob/5e1f551d86fcb5b731e3a83c0b043dbd7ef3c7aa/met/src/tools/other/pb2nc/dumppb.f#L144) for a description of the header columns you were asking about.

However, these are NOT observations of 6-hour precip... only state variables.

Next, I ran re-ran PB2NC using:
```
obs_var = [ "TP06", "TP12", "TP24" ];
```

And I got 10,608 total precip observations in the output. Most values were 0 although some were -0.1 (no idea why!). Perhaps -0.1 indicates bad data? I dumped the output NetCDF file to ascii to make is easier to read:
```
Rscript ../scripts/Rscripts/pntnc2ascii.R tp_obs.nc > tp_obs.txt
head -5 tp_obs.txt
```

Here's some sample data:
```
ADPSFC 03302 20120410_115000 53.25   -4.53   11 TP06 986.5   11.027688 NA  0.0 
ADPSFC 03354 20120410_115000 53.00   -1.25  117 TP06 973.700012207031  117.212425 NA  0.0 
ADPSFC 03502 20120410_115000 52.13   -4.57  134 TP06 974  133.706192 NA -0.1 
ADPSFC 78858 20120410_120000 17.30  -62.68   48 TP12 1009.59997558594   47.873726 NA  0.4 
ADPSFC 07002 20120410_120000 50.73    1.60   74 TP06 984.799987792969   74.117126 NA  0.0 
```

So plenty of good data. But your question is, how can you further filter these observations down by in/report/instrument type?

Rerunning my command at verbosity 10 yielded some clues in the following types of log messages. I wrote the log message to a file named run_pb2nc.log and grepped from that:
```
grep -B 1 TP06 run_pb2nc.log  | grep add_header_prepbufr | sort -u
```

Which yields:
```
DEBUG 9: add_header_prepbufr() pb_report_type: 181, in_report_type: 511, instrument_type: -9999
DEBUG 9: add_header_prepbufr() pb_report_type: 181, in_report_type: 512, instrument_type: -9999
DEBUG 9: add_header_prepbufr() pb_report_type: 183, in_report_type: 511, instrument_type: -9999
DEBUG 9: add_header_prepbufr() pb_report_type: 183, in_report_type: 512, instrument_type: -9999
DEBUG 9: add_header_prepbufr() pb_report_type: 187, in_report_type: 512, instrument_type: -9999
```

I'm guessing that pb_report_type values of 181, 183, and 187 and in_report_types of 511 and 512 correspond to TP06 observations.

More searching leads to this explanation:
https://www.emc.ncep.noaa.gov/mmb/data_processing/on124.htm#Tb_SM1

```
                   | Land station
           511     |     By block and station number (international
                   |       index number)
           512     |     By call letters
           513     |     By latitude-longitude
```

So the in_report_type values of 511 and 512 do NOT look very promising.

Looking at this link, I do see a mention of values 181, 183, and 187:
https://www.emc.ncep.noaa.gov/mmb/data_processing/prepbufr.doc/table_4.htm

However, I do not see a way to differentiate between DOT and non-DOT precip observations. I'd recommend asking for support from NOAA/EMC on these details.

**Comment by craigrj** [comment]
Thanks John, this is much more information than I could find from googling.  That gives me a place to start.  Since observations are the back bone of verification, any description of various types would be very helpful.  

 Bob

**Comment by JohnHalleyGotway** [comment]
Bob, I'm not sure what data source you're actually using. GDAS, NDAS, something else?
But I wanted to point out one detail. The PrepBUFR files are the results of some data processing that done at NOAA/EMC. For example, take a look at this ftp site containing recent GDAS data:
https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gdas.20220202/00/atmos/

As I understand it, the file [gdas.t00z.prepbufr.nr](https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gdas.20220202/00/atmos/gdas.t00z.prepbufr.nr) contains ALL of the non-restricted (nr) observations used in the GDAS data assimilation system. And we typically also use these observations for verification.

However that directory also contains lots of other files that include "bufr" in their name. I suspect these are the BUFR files that are INPUT into the GDAS data assimilation system. Now, I haven't actually run these through PB2NC myself, but I believe they should work. But I'd definitely encourage you to test that out!

It may be the case that processing these individual BUFR files would give you more control over the observations than the combined PrepBUFR files do. I'm not the right person to explain the differences between those BUFR files, but presumably someone at NOAA/EMC could.

Something to consider.

**Comment by craigrj** [comment]
Thanks, I sent an email to NOAA/EMC.  Do you guys have a good point of contact their?  I will give your idea a try.

 

Bob

**Comment by JohnHalleyGotway** [comment]
Bob, no I don’t have a specific contact in the DA group to recommend. Perhaps @PerryShafran-NOAA or @malloryprow would know who to ask questions about EMC data assimilation?

One other concern comes to mind. You are running an old version of MET which may NOT be able to process those separate BUFR files. I do not KNOW that to be the case, but if your pb2nc run fails please let me know. I can try running it on a newer version to see if the behavior differs.

**Comment by craigrj** [comment]
Alas, it seems to be a mountain too high for use to upgrade MET from 8.1.  We used to do that ourselves but now contractors have to do it and red tape gets in the way.

Thanks

Bob

**Comment by PerryShafran-NOAA** [comment]
I would suggest that rather than checking with the DA group, that you should speak to the obsproc group at EMC, as they are the group who processes the bufr dumps to create the prepbufr files.  Contact obsproc_support@noaa.gov, and they should be able to either answer the question or put you in contact with the person who would know about the bufr headings.

**Comment by JohnHalleyGotway** [comment]
@craigrj, wondering if you are seeking any more input on this discussion. Or should we mark it as being answered? The METplus team monitors open discussions daily, so we try get them wrapped up when possible.

**Comment by craigrj** [comment]
John, One thing I did learn from this issue which might be helpful to others is that when using observations for precipitation verification, set the observation time limits in the point stat config file to -10 to 0.  I found that many stations had precipitation reports in prepbufr with precip indicated in the 10 minute period before the hour and then reporting 0 precipitation within 5 minutes after the hour.  Originally I was pulling observations for +30,-30 minutes around valid time.  But once I changed that to -10,0 around valid time, skill scores improved significantly.

MET Lesson learned.

Bob

**Reply by JohnHalleyGotway** [reply]
Thanks for the feedback Bob. Yes, that is a good tip. I'll go ahead and mark this discussion as being answered.
